File: src/ghom/objects/Team.ts
// Team
// A team of users in an organization.
//
// Implements
// Node
// Subscribable
// Connections
// ancestors (TeamConnection!)
// A list of teams that are ancestors of this team.
//
// Argument	Type	Description
// after	String
// Returns the elements in the list that come after the specified cursor.
//
// before	String
// Returns the elements in the list that come before the specified cursor.
//
// first	Int
// Returns the first n elements from the list.
//
// last	Int
// Returns the last n elements from the list.
//
// childTeams (TeamConnection!)
// List of child teams belonging to this team
//
// Argument	Type	Description
// after	String
// Returns the elements in the list that come after the specified cursor.
//
// before	String
// Returns the elements in the list that come before the specified cursor.
//
// first	Int
// Returns the first n elements from the list.
//
// immediateOnly	Boolean
// Whether to list immediate child teams or all descendant child teams.
//
// The default value is true.
//
// last	Int
// Returns the last n elements from the list.
//
// orderBy	TeamOrder
// Order for connection
//
// userLogins	[String!]
// User logins to filter by
//
// discussions (TeamDiscussionConnection!)
// This part of the schema is currently available for developers to preview. During this preview period, the API may change without any advance notice. Please see the Team discussions preview for more details.
//
// Note: The GraphQL resources under preview cannot be accessed via the Explorer at this time.
//
// A list of team discussions.
//
// Argument	Type	Description
// after	String
// Returns the elements in the list that come after the specified cursor.
//
// before	String
// Returns the elements in the list that come before the specified cursor.
//
// first	Int
// Returns the first n elements from the list.
//
// isPinned	Boolean
// If provided, filters discussions according to whether or not they are pinned.
//
// last	Int
// Returns the last n elements from the list.
//
// orderBy	TeamDiscussionOrder
// Order for connection
//
// invitations (OrganizationInvitationConnection)
// A list of pending invitations for users to this team
//
// Argument	Type	Description
// after	String
// Returns the elements in the list that come after the specified cursor.
//
// before	String
// Returns the elements in the list that come before the specified cursor.
//
// first	Int
// Returns the first n elements from the list.
//
// last	Int
// Returns the last n elements from the list.
//
// members (TeamMemberConnection!)
// A list of users who are members of this team.
//
// Argument	Type	Description
// after	String
// Returns the elements in the list that come after the specified cursor.
//
// before	String
// Returns the elements in the list that come before the specified cursor.
//
// first	Int
// Returns the first n elements from the list.
//
// last	Int
// Returns the last n elements from the list.
//
// membership	TeamMembershipType
// Filter by membership type
//
// The default value is ALL.
//
// orderBy	TeamMemberOrder
// Order for the connection.
//
// query	String
// The search string to look for.
//
// role	TeamMemberRole
// Filter by team member role
//
// repositories (TeamRepositoryConnection!)
// A list of repositories this team has access to.
//
// Argument	Type	Description
// after	String
// Returns the elements in the list that come after the specified cursor.
//
// before	String
// Returns the elements in the list that come before the specified cursor.
//
// first	Int
// Returns the first n elements from the list.
//
// last	Int
// Returns the last n elements from the list.
//
// orderBy	TeamRepositoryOrder
// Order for the connection.
//
// query	String
// The search string to look for.
//
// Fields
// avatarUrl (URI)
// A URL pointing to the team's avatar.
//
// Argument	Type	Description
// size	Int
// The size in pixels of the resulting square image.
//
// The default value is 400.
//
// combinedSlug (String!)
// The slug corresponding to the organization and team.
//
// createdAt (DateTime!)
// Identifies the date and time when the object was created.
//
// description (String)
// The description of the team.
//
// discussion (TeamDiscussion)
// This part of the schema is currently available for developers to preview. During this preview period, the API may change without any advance notice. Please see the Team discussions preview for more details.
//
// Note: The GraphQL resources under preview cannot be accessed via the Explorer at this time.
//
// Find a team discussion by its number.
//
// Argument	Type	Description
// number	Int!
// The sequence number of the discussion to find.
//
// discussionsResourcePath (URI!)
// This part of the schema is currently available for developers to preview. During this preview period, the API may change without any advance notice. Please see the Team discussions preview for more details.
//
// Note: The GraphQL resources under preview cannot be accessed via the Explorer at this time.
//
// The HTTP path for team discussions
//
// discussionsUrl (URI!)
// This part of the schema is currently available for developers to preview. During this preview period, the API may change without any advance notice. Please see the Team discussions preview for more details.
//
// Note: The GraphQL resources under preview cannot be accessed via the Explorer at this time.
//
// The HTTP URL for team discussions
//
// editTeamResourcePath (URI!)
// The HTTP path for editing this team
//
// editTeamUrl (URI!)
// The HTTP URL for editing this team
//
// id (ID!)
// membersResourcePath (URI!)
// The HTTP path for the team' members
//
// membersUrl (URI!)
// The HTTP URL for the team' members
//
// name (String!)
// The name of the team.
//
// newTeamResourcePath (URI!)
// The HTTP path creating a new team
//
// newTeamUrl (URI!)
// The HTTP URL creating a new team
//
// organization (Organization!)
// The organization that owns this team.
//
// parentTeam (Team)
// The parent team of the team.
//
// privacy (TeamPrivacy!)
// The level of privacy the team has.
//
// repositoriesResourcePath (URI!)
// The HTTP path for this team's repositories
//
// repositoriesUrl (URI!)
// The HTTP URL for this team's repositories
//
// resourcePath (URI!)
// The HTTP path for this team
//
// slug (String!)
// The slug corresponding to the team.
//
// teamsResourcePath (URI!)
// The HTTP path for this team's teams
//
// teamsUrl (URI!)
// The HTTP URL for this team's teams
//
// updatedAt (DateTime!)
// Identifies the date and time when the object was last updated.
//
// url (URI!)
// The HTTP URL for this team
//
// viewerCanAdminister (Boolean!)
// Team is adminable by the viewer.
//
// viewerCanSubscribe (Boolean!)
// Check if the viewer is able to change their subscription status for the repository.
//
// viewerSubscription (SubscriptionState)
// Identifies if the viewer is watching, not watching, or ignoring the subscribable entity.

File: TypeScript/resources/code/GitHub/oktokit/octokit.graphql.net/Octokit.GraphQL/Model/Team.cs
namespace Octokit.GraphQL.Model
{
    using System;
    using System.Collections.Generic;
    using System.Linq.Expressions;
    using Octokit.GraphQL.Core;
    using Octokit.GraphQL.Core.Builders;

    /// <summary>
    /// A team of users in an organization.
    /// </summary>
    public class Team : QueryableValue<Team>
    {
        internal Team(Expression expression) : base(expression)
        {
        }

        /// <summary>
        /// A list of teams that are ancestors of this team.
        /// </summary>
        /// <param name="first">Returns the first _n_ elements from the list.</param>
        /// <param name="after">Returns the elements in the list that come after the specified cursor.</param>
        /// <param name="last">Returns the last _n_ elements from the list.</param>
        /// <param name="before">Returns the elements in the list that come before the specified cursor.</param>
        public TeamConnection Ancestors(Arg<int>? first = null, Arg<string>? after = null, Arg<int>? last = null, Arg<string>? before = null) => this.CreateMethodCall(x => x.Ancestors(first, after, last, before), Octokit.GraphQL.Model.TeamConnection.Create);

        /// <summary>
        /// A URL pointing to the team's avatar.
        /// </summary>
        /// <param name="size">The size in pixels of the resulting square image.</param>
        public string AvatarUrl(Arg<int>? size = null) => default;

        /// <summary>
        /// List of child teams belonging to this team
        /// </summary>
        /// <param name="first">Returns the first _n_ elements from the list.</param>
        /// <param name="after">Returns the elements in the list that come after the specified cursor.</param>
        /// <param name="last">Returns the last _n_ elements from the list.</param>
        /// <param name="before">Returns the elements in the list that come before the specified cursor.</param>
        /// <param name="immediateOnly">Whether to list immediate child teams or all descendant child teams.</param>
        /// <param name="orderBy">Order for connection</param>
        /// <param name="userLogins">User logins to filter by</param>
        public TeamConnection ChildTeams(Arg<int>? first = null, Arg<string>? after = null, Arg<int>? last = null, Arg<string>? before = null, Arg<bool>? immediateOnly = null, Arg<TeamOrder>? orderBy = null, Arg<IEnumerable<string>>? userLogins = null) => this.CreateMethodCall(x => x.ChildTeams(first, after, last, before, immediateOnly, orderBy, userLogins), Octokit.GraphQL.Model.TeamConnection.Create);

        /// <summary>
        /// The slug corresponding to the organization and team.
        /// </summary>
        public string CombinedSlug { get; }

        /// <summary>
        /// Identifies the date and time when the object was created.
        /// </summary>
        public DateTimeOffset CreatedAt { get; }

        /// <summary>
        /// Identifies the primary key from the database.
        /// </summary>
        public int? DatabaseId { get; }

        /// <summary>
        /// The description of the team.
        /// </summary>
        public string Description { get; }

        /// <summary>
        /// Find a team discussion by its number.
        /// </summary>
        /// <param name="number">The sequence number of the discussion to find.</param>
        public TeamDiscussion Discussion(Arg<int> number) => this.CreateMethodCall(x => x.Discussion(number), Octokit.GraphQL.Model.TeamDiscussion.Create);

        /// <summary>
        /// A list of team discussions.
        /// </summary>
        /// <param name="first">Returns the first _n_ elements from the list.</param>
        /// <param name="after">Returns the elements in the list that come after the specified cursor.</param>
        /// <param name="last">Returns the last _n_ elements from the list.</param>
        /// <param name="before">Returns the elements in the list that come before the specified cursor.</param>
        /// <param name="isPinned">If provided, filters discussions according to whether or not they are pinned.</param>
        /// <param name="orderBy">Order for connection</param>
        public TeamDiscussionConnection Discussions(Arg<int>? first = null, Arg<string>? after = null, Arg<int>? last = null, Arg<string>? before = null, Arg<bool>? isPinned = null, Arg<TeamDiscussionOrder>? orderBy = null) => this.CreateMethodCall(x => x.Discussions(first, after, last, before, isPinned, orderBy), Octokit.GraphQL.Model.TeamDiscussionConnection.Create);

        /// <summary>
        /// The HTTP path for team discussions
        /// </summary>
        public string DiscussionsResourcePath { get; }

        /// <summary>
        /// The HTTP URL for team discussions
        /// </summary>
        public string DiscussionsUrl { get; }

        /// <summary>
        /// The HTTP path for editing this team
        /// </summary>
        public string EditTeamResourcePath { get; }

        /// <summary>
        /// The HTTP URL for editing this team
        /// </summary>
        public string EditTeamUrl { get; }

        public ID Id { get; }

        /// <summary>
        /// A list of pending invitations for users to this team
        /// </summary>
        /// <param name="first">Returns the first _n_ elements from the list.</param>
        /// <param name="after">Returns the elements in the list that come after the specified cursor.</param>
        /// <param name="last">Returns the last _n_ elements from the list.</param>
        /// <param name="before">Returns the elements in the list that come before the specified cursor.</param>
        public OrganizationInvitationConnection Invitations(Arg<int>? first = null, Arg<string>? after = null, Arg<int>? last = null, Arg<string>? before = null) => this.CreateMethodCall(x => x.Invitations(first, after, last, before), Octokit.GraphQL.Model.OrganizationInvitationConnection.Create);

        /// <summary>
        /// Get the status messages members of this entity have set that are either public or visible only to the organization.
        /// </summary>
        /// <param name="first">Returns the first _n_ elements from the list.</param>
        /// <param name="after">Returns the elements in the list that come after the specified cursor.</param>
        /// <param name="last">Returns the last _n_ elements from the list.</param>
        /// <param name="before">Returns the elements in the list that come before the specified cursor.</param>
        /// <param name="orderBy">Ordering options for user statuses returned from the connection.</param>
        public UserStatusConnection MemberStatuses(Arg<int>? first = null, Arg<string>? after = null, Arg<int>? last = null, Arg<string>? before = null, Arg<UserStatusOrder>? orderBy = null) => this.CreateMethodCall(x => x.MemberStatuses(first, after, last, before, orderBy), Octokit.GraphQL.Model.UserStatusConnection.Create);

        /// <summary>
        /// A list of users who are members of this team.
        /// </summary>
        /// <param name="first">Returns the first _n_ elements from the list.</param>
        /// <param name="after">Returns the elements in the list that come after the specified cursor.</param>
        /// <param name="last">Returns the last _n_ elements from the list.</param>
        /// <param name="before">Returns the elements in the list that come before the specified cursor.</param>
        /// <param name="membership">Filter by membership type</param>
        /// <param name="orderBy">Order for the connection.</param>
        /// <param name="query">The search string to look for.</param>
        /// <param name="role">Filter by team member role</param>
        public TeamMemberConnection Members(Arg<int>? first = null, Arg<string>? after = null, Arg<int>? last = null, Arg<string>? before = null, Arg<TeamMembershipType>? membership = null, Arg<TeamMemberOrder>? orderBy = null, Arg<string>? query = null, Arg<TeamMemberRole>? role = null) => this.CreateMethodCall(x => x.Members(first, after, last, before, membership, orderBy, query, role), Octokit.GraphQL.Model.TeamMemberConnection.Create);

        /// <summary>
        /// The HTTP path for the team' members
        /// </summary>
        public string MembersResourcePath { get; }

        /// <summary>
        /// The HTTP URL for the team' members
        /// </summary>
        public string MembersUrl { get; }

        /// <summary>
        /// The name of the team.
        /// </summary>
        public string Name { get; }

        /// <summary>
        /// The HTTP path creating a new team
        /// </summary>
        public string NewTeamResourcePath { get; }

        /// <summary>
        /// The HTTP URL creating a new team
        /// </summary>
        public string NewTeamUrl { get; }

        /// <summary>
        /// The organization that owns this team.
        /// </summary>
        public Organization Organization => this.CreateProperty(x => x.Organization, Octokit.GraphQL.Model.Organization.Create);

        /// <summary>
        /// The parent team of the team.
        /// </summary>
        public Team ParentTeam => this.CreateProperty(x => x.ParentTeam, Octokit.GraphQL.Model.Team.Create);

        /// <summary>
        /// The level of privacy the team has.
        /// </summary>
        public TeamPrivacy Privacy { get; }

        /// <summary>
        /// A list of repositories this team has access to.
        /// </summary>
        /// <param name="first">Returns the first _n_ elements from the list.</param>
        /// <param name="after">Returns the elements in the list that come after the specified cursor.</param>
        /// <param name="last">Returns the last _n_ elements from the list.</param>
        /// <param name="before">Returns the elements in the list that come before the specified cursor.</param>
        /// <param name="orderBy">Order for the connection.</param>
        /// <param name="query">The search string to look for.</param>
        public TeamRepositoryConnection Repositories(Arg<int>? first = null, Arg<string>? after = null, Arg<int>? last = null, Arg<string>? before = null, Arg<TeamRepositoryOrder>? orderBy = null, Arg<string>? query = null) => this.CreateMethodCall(x => x.Repositories(first, after, last, before, orderBy, query), Octokit.GraphQL.Model.TeamRepositoryConnection.Create);

        /// <summary>
        /// The HTTP path for this team's repositories
        /// </summary>
        public string RepositoriesResourcePath { get; }

        /// <summary>
        /// The HTTP URL for this team's repositories
        /// </summary>
        public string RepositoriesUrl { get; }

        /// <summary>
        /// The HTTP path for this team
        /// </summary>
        public string ResourcePath { get; }

        /// <summary>
        /// The slug corresponding to the team.
        /// </summary>
        public string Slug { get; }

        /// <summary>
        /// The HTTP path for this team's teams
        /// </summary>
        public string TeamsResourcePath { get; }

        /// <summary>
        /// The HTTP URL for this team's teams
        /// </summary>
        public string TeamsUrl { get; }

        /// <summary>
        /// Identifies the date and time when the object was last updated.
        /// </summary>
        public DateTimeOffset UpdatedAt { get; }

        /// <summary>
        /// The HTTP URL for this team
        /// </summary>
        public string Url { get; }

        /// <summary>
        /// Team is adminable by the viewer.
        /// </summary>
        public bool ViewerCanAdminister { get; }

        /// <summary>
        /// Check if the viewer is able to change their subscription status for the repository.
        /// </summary>
        public bool ViewerCanSubscribe { get; }

        /// <summary>
        /// Identifies if the viewer is watching, not watching, or ignoring the subscribable entity.
        /// </summary>
        public SubscriptionState? ViewerSubscription { get; }

        internal static Team Create(Expression expression)
        {
            return new Team(expression);
        }
    }
}